### ISSUE TITLE
Clang does not detect misaligned store

### ISSUE THREAD

**Issue by vitalybuka** [llvm/llvm-project#199785]
https://godbolt.org/z/z6r5vqe4s

```c++
struct A {
    int a;
    int b;
    double c;
};

struct B {
    char c;
    A a[10];
} __attribute__((packed));

alignas(32) B b[2];
A* p = &b[1].a[5];

A v;

int main() {
    *p = {
        .a = 3,
        .b = 5,
        .c = 7,
    };
}
```

It still produces an aligned store in IR.

```llvm
define dso_local noundef i32 @main() #0 !func_sanitize !39 {
entry:
  %ref.tmp = alloca %struct.A, align 8
  %a = getelementptr inbounds nuw %struct.A, ptr %ref.tmp, i32 0, i32 0
  store i32 3, ptr %a, align 8
  %b = getelementptr inbounds nuw %struct.A, ptr %ref.tmp, i32 0, i32 1
  store i32 5, ptr %b, align 4
  %c = getelementptr inbounds nuw %struct.A, ptr %ref.tmp, i32 0, i32 2
  store double 7.000000e+00, ptr %c, align 8
  %0 = load ptr, ptr @p, align 8
  call void @llvm.memcpy.p0.p0.i64(ptr align 8 %0, ptr align 8 %ref.tmp, i64 16, i1 false)
  ret i32 0
}
```

**Comment by efriedma-quic** [comment]
It looks like #190739 fixed this for the RHS of assignment operators, but not the LHS.  CC @vasu-the-sharma.

**Comment by vitalybuka** [comment]
Yep, I started an investigation of new reports after #190739.

It detects in C https://godbolt.org/z/sPhMqhz5f, but not C++ above.

**Comment by corbettknowles** [comment]
Traced to emitPointerArithmetic in CGExpr.cpp. The problem is on line 1454 — getArrayElementAlign(BaseAddr.getAlignment(), index, ...). On the LHS path BaseAddr.getAlignment() does not reflect the packed struct alignment reduction. The fix from #190739 addressed the RHS but this LHS path through emitPointerArithmetic is missing the same packed alignment cap before the return Address on line 1457 that file in codegen folder CGExpr.cpp
